File: Stacks/largest_rectangle_in_histogram.py
"""
https://leetcode.com/problems/largest-rectangle-in-histogram/description/
"""


class Solution: #Time complexity O(n), Space complexity O(n)
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        res = 0

        for index, height in enumerate(heights):
            i = index
            while stack and stack[-1][1] > height:
                i, h = stack.pop()
                res = max(res, h * (index - i))
            
            stack.append((i, height))

        index = len(heights)
        
        while stack:
            i, h = stack.pop()
            res = max(res, h * (index - i))

        return res


        

File: Stacks/largest_rectangle_in_histogram.js
/*
https://leetcode.com/problems/largest-rectangle-in-histogram/description/
*/

class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     * Time complexity O(n), Space complexity O(n)
     */
    largestRectangleArea(heights) {
        const stack = []
        var res = 0

        for(let i = 0; i < heights.length; i++){
            let start = i
            while(stack.length && stack[stack.length-1][1] > heights[i]){
                var [index, height] = stack.pop()
                res = Math.max(res, height * (i - index))
                start = index
            }
            stack.push([start, heights[i]])
        }

        const i = heights.length
        while(stack.length){
            var [index, height] = stack.pop()
            res = Math.max(res, height * (i - index))  
        }

        return res
    }
}
